File: exercises/java/002_adapted_response_pattern/src/main/java/com/akka/training/BaristaActor.java
package com.akka.training;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class BaristaActor extends AbstractBehavior<BaristaActor.BaristaCommand> {
  // Orders <Whom, Coffee>
  private final Map<String, Coffee> orders = new HashMap<>();
  // reference to the coffee-machine child actor, allowing to send messages to coffee machine
  private final ActorRef<CoffeeMachineActor.CoffeeMachineCommand> coffeeMachine;

  private BaristaActor(ActorContext<BaristaCommand> context) {
    super(context);
    // We spawn the CoffeeMachine as child actor in the private constructor where we have access to
    // both context and class fields.
    // It returns an ActorRef that we need to keep in the state, we will need it to interact with
    // the CoffeeMachine.
    coffeeMachine = context.spawn(CoffeeMachineActor.create(), "coffee-machine");
  }

  public static Behavior<BaristaCommand> create() {
    return Behaviors.setup(BaristaActor::new);
  }

  // Format the orders into expected format [whom1->coffee1,whom2->coffee2]
  public static String printOrders(Set<Map.Entry<String, Coffee>> orders) {
    return orders.stream()
        .map(kv -> String.format("%s->%s", kv.getKey(), kv.getValue()))
        .reduce((acc, s) -> acc + "," + s)
        .map(s -> "[" + s + "]")
        .orElse("[" + "]");
  }

  @Override
  public Receive<BaristaCommand> createReceive() {
    return newReceiveBuilder().onMessage(OrderCoffee.class, this::onOrderCoffee).build();
  }

  private Behavior<BaristaCommand> onOrderCoffee(OrderCoffee command) {
    orders.put(command.whom, command.coffee);
    getContext().getLog().info("Orders: {}", printOrders(orders.entrySet()));

    return this;
  }

  // <- Protocol definition
  public sealed interface BaristaCommand permits OrderCoffee {}

  public record OrderCoffee(String whom, Coffee coffee) implements BaristaCommand {}
  // Protocol definition ->
}

File: exercises/scala/002_adapted-response-pattern/src/main/scala/dev/fResult/akka/BaristaActor.scala
package dev.fResult.akka

import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.{AbstractBehavior, ActorContext, Behaviors}

import scala.collection.mutable

object BaristaActor:
  def apply(): Behavior[OrderCoffee] = Behaviors.setup(BaristaBehavior(_))

  private def printOrder(orderSet: Set[(String, Coffee)]): String = {
    val formattedOrders = orderSet.map(order => s"${order._1}->${order._2}")

    s"[${formattedOrders.mkString(", ")}]"
  }

  final case class OrderCoffee(whom: String, coffee: Coffee)

  private class BaristaBehavior(context: ActorContext[OrderCoffee]) extends AbstractBehavior[OrderCoffee](context):
    private val orders: mutable.Map[String, Coffee] = mutable.Map()

    override def onMessage(message: OrderCoffee): Behavior[OrderCoffee] = {
      orders.put(message.whom, message.coffee)

      context.log.info(s"Orders: ${printOrder(orders.toSet)}")

      Behaviors.same
    }
  end BaristaBehavior
end BaristaActor
